File: Part 2/Chapter 7/quick_sort.py
def partition1(array:list, low, high) -> int:
    pivot = array[high]
    i = low - 1
    for j in range(low, high):
        if array[j] < pivot:
            i += 1
            array[i], array[j] = array[j], array[i]
    array[i + 1], array[high] = array[high], array[i + 1]
    return i + 1

def quick_sort1(array:list, low, high):
    if low < high:
        mid = partition1(array, low, high)
        quick_sort1(array, low, mid - 1)
        quick_sort1(array, mid + 1, high)

def partition2(array:list, low, high) -> int:
    pivot = array[low]
    l = low
    r = high
    while l < r:
        while l < r and array[r] >= pivot:
            r -= 1
        array[l] = array[r]

        while l < r and array[l] <= pivot:
            l += 1
        array[r] = array[l]

    array[l] = pivot
    return l

def quick_sort2(array:list, low, high):
    if low < high:
        mid = partition2(array, low, high)
        quick_sort2(array, low, mid - 1)
        quick_sort2(array, mid + 1, high)


import time
import numpy as np

a = np.random.randint(0, 1000, 100000)
b = a.copy()

t = time.process_time()
quick_sort1(a, 0, len(a) - 1)
print(time.process_time() - t)

t = time.process_time()
quick_sort2(b, 0, len(b) - 1)
print(time.process_time() - t)

print(np.allclose(a, b))

File: Part 2/Chapter 7/Quick Sort.cpp
#include <iostream>
#include <random>
#include <vector>

using namespace std;

template <typename T>
int partition(vector<T>& a, int low, int high)
{
    T pivot = a[high - 1];
    int i = low - 1;
    for (int j = low; j < high; j++)
    {
        if (a[j] < pivot)
        {
            i++;
            swap(a[i], a[j]);
        }
    }
    swap(a[i + 1], a[high - 1]);
    return i + 1;
}

template <typename T>
void QSort(vector<T>& a, int low, int high)
{
    if (low < high)
    {
        int mid = partition(a, low, high);
        QSort(a, low, mid);
        QSort(a, mid + 1, high);
    }
}

int main()
{
    vector<int> data;
    int n = 121;
    for (int i = 0; i < n; i++)
    {
        data.push_back(i);
    }
    shuffle(data.begin(), data.end(), default_random_engine(0));
    for(auto& i : data)
    {
        cout << i << " ";
    }
    cout << endl;

    QSort(data, 0, n);

    for(auto& i : data)
    {
        cout << i << " ";
    }
    cout << endl;
    return 0;
}